sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala

```diff
@@ -561,14 +566,18 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] with AliasHelpe
         val origOutput = query.output.head
 
         val resultWithZeroTups = evalSubqueryOnZeroTups(query)
-        if (resultWithZeroTups.isEmpty) {
+        lazy val (topPart, havingNode, aggNode) = splitSubquery(query)
```

Comment:
```suggestion
        val (topPart, havingNode, aggNode) = splitSubquery(query)
```

Reply:
`aggNode` is accessed immediately.

Reply:
Wouldn't this short circuit?
```
if (resultWithZeroTups.isEmpty || aggNode.isEmpty) {
```
If `resultWithZeroTups` is empty, then `aggNode` won't get accessed. I am trying to prevent splitting the subquery when it is not subject to the count bug.